lib/GoalCohort.js

```diff
@@ -122,22 +144,32 @@ GoalCohort.statics.findUserCohort = async function(user, goal) {
  *		adds the user to that cohort and returns it
  * 3) If neither of the above, creates a new GoalCohort
  *		and adds the user as its first member
+ * 4) If the user is in a user cohort, join or create a GoalCohort with that name
+ * 5) If the user is not in a user cohort, join or create a GoalCohort without a name
  */
 GoalCohort.statics.joinOrCreateCohort = async function(user, goal, opts) {
   const goalCohortModel = this;
+  const userCohortModel = mongoose.model("UserCohort");
+  const userCohort = await userCohortModel.findForUser(user);
+  const userCohortName = userCohort ? userCohort.cohort : undefined;
 
-  let cohort = await goalCohortModel.findOne({
-    goal: goal.id,
-    members: {
-      $elemMatch: {
-        user: user.id
-      }
+  let cohort = await goalCohortModel.findUserCohort(user, goal);
+  if (cohort) {
+    if (cohort.nameCanonical === canonicalize(userCohortName)) {
+      return cohort;
     }
-  });
+    cohort = await goalCohortModel.leaveCohort(user, goal);
+  }
 
-  // If the user already belongs to a cohort for the given goal, return it
-  if (cohort) {
-    return cohort;
+  if (userCohort) {
+    cohort = await goalCohortModel.findCohortByName(userCohortName, goal);
+    if (cohort && cohort.memberSlotsRemaining === 0) {
+      const err = new Error(
```

Comment:
What is the scenario for this from the user's perspective? Is there a reason we should not just auto create the team in this case (especially since we would auto-create the goal cohort)?

Whichever we decide is the correct/exptect behavior, we should have a test for it too, e.g. if we *do* want the auto-team create behavior, then there should be at least a test like

```
describe("joinOrCreateCohort", () => {
  it("creates a new team when user belongs to a user cohort and no space on existing teams", 
```

Reply:
That's when the user is trying to auto-join any available team. But in this case, the named cohort is full (does not have any available teams). If we auto create a team, what would the name be? The icon? Just pick a random team name and icon? Would the user even want to create a team seeing as they specifically requested the auto join option?

Reply:
Do we not autogenerate team names in other places? My sense is when a user clicks "Auto Join Team" they just want to be on a team in their cohort and they don't care about they name.